[PFC] Native UCC process group for Pytorch

vtlam opened the pull request:
Summary:
This diff integrates UCC process group as a native component of Pytorch Distributed core. It is based on the existing torch-ucc (https://github.com/facebookresearch/torch_ucc) as the wrapper for UCC collective communication library.
The environment and cmake variables are named in mirroring to the existing process groups such as NCCL and Gloo. Specifically,
- USE_UCC: enables UCC PG. This defaults to OFF, so there is no breakage of existing builds that do not have UCX/UCC external libraries.
- USE_SYSTEM_UCC: uses external UCX and UCC shared libraries that are set accordingly with UCX_HOME and UCC_HOME.

Currently, this diff only supports USE_SYSTEM_UCC=ON, i.e., requiring users to specify external libraries for UCX and UCC. In subsequent diffs, we will add UCX and UCC repos as third-party dependencies in pytorch/third-party.

Test Plan:
Passed Torch-UCC tests that invoke UCC process group. For example:

$ sh test/start_test.sh test/torch_allreduce_test.py --backend gloo --use-cuda
...
Test allreduce: succeeded

Differential Revision: D36973688

zasdfgbnm commented on cmake/External/ucc.cmake:
```diff
@@ -0,0 +1,20 @@
+  set(__UCC_INCLUDED TRUE)
+
+  if(USE_SYSTEM_UCC)
+    set(UCX_HOME $ENV{UCX_HOME} CACHE PATH "UCX install directory")
+    set(UCC_HOME $ENV{UCC_HOME} CACHE PATH "UCC install directory")
```
Should we use [`find_package`](https://cmake.org/cmake/help/latest/command/find_package.html) to find the path for UCC and UCX? I added cmake support to UCX and UCC in the past:
https://github.com/openucx/ucx/pull/7096
https://github.com/openucx/ucc/pull/255

vtlam replied:
This PR only supports USE_SYSTEM_UCC for simplicity. I have a follow up PR on making UCC/UCX repos part of pytorch/third-party (and automatically compile them with pytorch cmake etc). I feel those are very cmake-specific and can be addressed as a follow up so that this PR can focus on UCC PG (which is already pretty big PR). I will look at your past PRs accordingly. Thanks for the pointers!

vtlam replied:
sgtm. thanks.

zasdfgbnm commented on cmake/Summary.cmake:
```diff
@@ -146,6 +146,10 @@ function(caffe2_print_configuration_summary)
     message(STATUS "  USE_MKLDNN_ACL        : ${USE_MKLDNN_ACL}")
     message(STATUS "  USE_MKLDNN_CBLAS      : ${USE_MKLDNN_CBLAS}")
   endif()
+  message(STATUS "  USE_UCC               : ${USE_UCC}")
+  if(${USE_UCC})
+    message(STATUS "  USE_SYSTEM_UCC        : ${USE_SYSTEM_UCC}")
+  endif()
```
```suggestion
  message(STATUS "  USE_UCC               : ${USE_UCC}")
  if(${USE_UCC})
    message(STATUS "     USE_SYSTEM_UCC        : ${USE_SYSTEM_UCC}")
  endif()
```

vtlam replied:
done

zasdfgbnm commented on torch/csrc/distributed/c10d/ProcessGroupUCC.cpp:
```diff
@@ -0,0 +1,1575 @@
+    {ReduceOp::BAND, UCC_OP_BAND},
+    {ReduceOp::BOR, UCC_OP_BOR},
+    {ReduceOp::BXOR, UCC_OP_BXOR},
+#if TORCH_VERSION_MAJOR > 1 || (TORCH_VERSION_MAJOR == 1 && TORCH_VERSION_MINOR >= 11)
```
I think this guard can be removed

vtlam replied:
done

zasdfgbnm commented on torch/csrc/distributed/c10d/ProcessGroupUCC.cpp:
```diff
@@ -0,0 +1,1575 @@
+      // bitwise and
+      return UCC_OP_MIN;
+    }
+#if TORCH_VERSION_MAJOR > 1 || \
+    (TORCH_VERSION_MAJOR == 1 && TORCH_VERSION_MINOR >= 11)
```
I think this guard can be removed

vtlam replied:
done

zasdfgbnm commented on torch/csrc/distributed/c10d/ProcessGroupUCC.cpp:
```diff
@@ -0,0 +1,1575 @@
+    }
+  }
+  torch_ucc_config.blocking_wait[(std::uint8_t)OpType::ALLGATHER] =
+      std::stoi(torch_ucc_envs_map.at("TORCH_UCC_ALLGATHER_BLOCKING_WAIT"));
```
Do we need a better name here? https://github.com/facebookresearch/torch_ucc/pull/77#discussion_r874102990

vtlam replied:
I suggest to keep these names as is so that we have an easy cross-reference to the old torch-ucc repo. Then let's discuss about naming and send subsequent PRs to rename.

zasdfgbnm commented on torch/csrc/distributed/c10d/ProcessGroupUCC.cpp:
```diff
@@ -0,0 +1,1575 @@
+}
+
+bool ProcessGroupUCC::WorkUCC::wait(std::chrono::milliseconds /* unused */) {
+#ifdef USE_CUDA
```
I think `USE_CUDA` is always true for now? But if you do want to use UCC on CPU-only builds in the future, we can keep this guard for now, so that in the future, we can easily figure out which code to move to `libtorch_cuda.so`.

vtlam replied:
let's keep it for CPU support as you mentioned.

zasdfgbnm commented on torch/distributed/distributed_c10d.py:
```diff
@@ -767,6 +785,32 @@ def _new_process_group_helper(
+                raise RuntimeError("Distributed package doesn't have UCC built in")
+
+            pg = ProcessGroupUCC(prefix_store, rank, world_size, timeout=timeout)
+            # In debug mode and if GLOO is available, wrap in a wrapper PG that
+            # enables enhanced collective checking for debugability.
+            if get_debug_level() == DebugLevel.DETAIL:
+                if not _GLOO_AVAILABLE:
+                    logger.info(
+                        """TORCH_DISTRIBUTED_DEBUG was set to DETAIL, but
+                                GLOO is not available. Build with Gloo to
+                                create a wrapper process group in debug mode
+                                to aid collective desynchronization debugging."""
+                    )
```
```suggestion
            # In debug mode and if GLOO is available, wrap in a wrapper PG that
            # enables enhanced collective checking for debugability.
            if get_debug_level() == DebugLevel.DETAIL:
                if not _UCC_AVAILABLE:
                    logger.info(
                        """TORCH_DISTRIBUTED_DEBUG was set to DETAIL, but
                                UCC is not available. Build with UCC to
                                create a wrapper process group in debug mode
                                to aid collective desynchronization debugging."""
                    )
```

vtlam replied:
I think Gloo is the correct one as that's a separate PG being used as a wrapper PG for debugging (see the same logic with NCCL PG right above these lines).

vtlam commented on caffe2/CMakeLists.txt:
```diff
@@ -968,6 +968,13 @@ elseif(USE_CUDA)
     target_link_libraries(torch_cuda PRIVATE __caffe2_nccl)
     target_compile_definitions(torch_cuda PRIVATE USE_NCCL)
   endif()
+  if(USE_UCC AND BUILD_SPLIT_CUDA)
+    target_link_libraries(torch_cuda_cpp PRIVATE __caffe2_ucc)
+    target_compile_definitions(torch_cuda_cpp PRIVATE USE_UCC)
+  elseif(USE_UCC)
+    target_link_libraries(torch_cuda PRIVATE __caffe2_ucc)
+    target_compile_definitions(torch_cuda PRIVATE USE_UCC)
+  endif()
```
i added dependency on torch_cpu as well per your suggestion. I imagine we'd like to support CPU-only build as well.

zasdfgbnm replied:
It doesn't have to be in this PR, but I think you are still building the UCC in libtorch_cuda, as indicated in https://github.com/pytorch/pytorch/pull/79918/files#diff-9f82c88fa2134b2830af1fe287b3d601c34ff481a8c989a86392bea7689b58abR744-R752
So you still can not use it in a CPU-only build? I think some bigger refactor is needed to allow CPU-only build, that is, to move all current code guarded with `#ifdef USE_CUDA` into a separate cpp, and put it in `libtorch_cuda`, and put the rest on `libtorch_cpu`. Feel free to leave this PR as is, and resolve this in later PR.

zasdfgbnm commented on build_variables.bzl:
```diff
@@ -744,6 +744,9 @@ libtorch_cuda_distributed_base_sources = [
 libtorch_cuda_distributed_extra_sources = [
     "torch/csrc/distributed/c10d/NCCLUtils.cpp",
     "torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp",
+    "torch/csrc/distributed/c10d/ProcessGroupUCC.cpp",
```
See: https://github.com/pytorch/pytorch/pull/79918/files#r906266685

vtlam replied:
Thanks for raising this issue! I got to understand your concern now. I agree that as it stands, UCC PG is not compiled when USE_CUDA=0. 
Nevertheless, for code readability I think it is simpler to have UCC implementation in one place  rather than breaking them into CPU-only and GPU pieces.
Can we instead solve the problem with the compiler configuration? 
For example, I have tried adding UCC files (e.g., ProcessGroupUCC.cpp, UCCUtils.cpp) to both libtorch_distributed_base_sources and libtorch_cuda_distributed_extra_sources.
Doing this seems to allow me to run UCC PG in both CPU-mode (compiling with USE_CUDA=0) and in GPU-mode (compiling with USE_CUDA=1). Let me know if you have other suggestions. Thanks.

zasdfgbnm replied:
I don't understand the build system of PyTorch very well, so I have no idea about what will happen if you do that. I am just wondering if you add them to both `libtorch_distributed_base_sources` and `libtorch_cuda_distributed_extra_sources`, is this effectively equal to copy-pasting the code twice? How does the linker know which to choose when encountering the same symbol?

vtlam replied:
I have a theory why it works, but I need to confirm on the compiler/linker workflow as Pytorch is kinda complex with CMake, Bazel etc.
In any case, I feel perhaps a more robust way is to perform explicit check on CUDA (e.g., with Bazel's if_cuda), and include the UCC files to the right file glob accordingly. I'd be happy as long as we can solve the problem with compiler config so that the UCC files can be kept clean.
As you mentioned, let's keep this PR as is and handle this issue later (I will try get advice from a Pytorch build expert).

malfet replied:
@zasdfgbnm linker should fail with conflicting symbols. Or, if none of the functions are marked as `TORCH_API` it will link it it twice and use twice the memory

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -1343,6 +1355,16 @@ if(USE_DISTRIBUTED)
     target_compile_definitions(torch_cpu PUBLIC USE_C10D_GLOO)
   endif()
+  if(USE_UCC AND USE_C10D_UCC)
+    target_compile_definitions(torch_cpu PUBLIC USE_C10D_UCC)
```
Why does it have to be a public definition? Should consumers of libtorch_cpu be aware about this one. If they do, we better put it in  the autogenerated TorchConfig.h

vtlam replied:
Mainly for consistency with other target_compile_definitions such as USE_C10D_GLOO, USE_C10D_NCCL, USE_C10D_MPI etc.

malfet commented on cmake/Summary.cmake:
```diff
@@ -146,6 +146,10 @@ function(caffe2_print_configuration_summary)
     message(STATUS "  USE_MKLDNN_CBLAS      : ${USE_MKLDNN_CBLAS}")
   endif()
+  message(STATUS "  USE_UCC               : ${USE_UCC}")
+  if(${USE_UCC})
+    message(STATUS "    USE_SYSTEM_UCC        : ${USE_SYSTEM_UCC}")
+  endif()
```
Why this is needed if `USE_SYSTEM_UCC=NO` is not yet supported?

malfet commented on torch/csrc/distributed/c10d/UCCTracing.cpp:
```diff
@@ -0,0 +1,176 @@
+    const int world_size,
+    const std::vector<at::Tensor>& inputTensors,
+    const std::vector<at::Tensor>& outputTensors) {
+  auto inSize = (inputTensors.size() > 0) ? inputTensors[0].numel() : 0;
```
Nit: Please use [`std::vector::empty`](https://en.cppreference.com/w/cpp/container/vector/empty) predicate
```suggestion
  auto inSize = (!inputTensors.empty()) ? inputTensors[0].numel() : 0;
```

malfet commented on torch/csrc/distributed/c10d/UCCUtils.cpp:
```diff
@@ -0,0 +1,274 @@
+}
+
+ucc_status_t oob_allgather_free(void* req) {
+  torch_ucc_oob_coll_info_t* info =
+      reinterpret_cast<torch_ucc_oob_coll_info_t*>(req);
```
Nit: Please use `auto`
```suggestion
  auto* info =
      reinterpret_cast<torch_ucc_oob_coll_info_t*>(req);
```

malfet replied:
Also,  don't you want to check that INFO is not empty?

malfet commented on torch/csrc/distributed/c10d/UCCUtils.cpp:
```diff
@@ -0,0 +1,274 @@
+    int num_done = info->store->add({info->getKey("ag_done")}, 1);
+    if (num_done == info->size) {
+      info->store->deleteKey(info->getKey("ag_done"));
+      for (int r = 0; r < info->size; r++) {
+        info->store->deleteKey(info->getKey("teamr" + std::to_string(r)));
+      }
+      for (int r = 0; r < info->size; r++) {
+        info->store->add({info->getKey("ag_free" + std::to_string(r))}, 1);
+      }
```
Is there a reason not to do it in the same loop? Also, consider using `c10::irange` primitive
```suggestion
      for (const auto r: c10::irange(info->size)) {
        info->store->deleteKey(info->getKey("teamr" + std::to_string(r)));
        info->store->add({info->getKey("ag_free" + std::to_string(r))}, 1);
      }
```

vtlam replied:
I added a comment that it is required as separate loop.

malfet commented on torch/csrc/distributed/c10d/UCCUtils.cpp:
```diff
@@ -0,0 +1,274 @@
+        info->store->add({info->getKey("ag_free" + std::to_string(r))}, 1);
+      }
+    } else {
+      info->store->wait({info->getKey("ag_free" + std::to_string(info->rank))});
```
Please refactor `ag_free` string literal in a constant, and use it throughout the code (otherwise its very easy to make a typo which will be hard to debug)

H-Huang commented on torch/csrc/distributed/c10d/ProcessGroupUCC.cpp:
```diff
@@ -0,0 +1,1870 @@
+  std::vector<uint8_t> remote_comm_id;
+  oob->store->deleteKey(group_id + std::to_string(0));
+  if (oob->rank != 0) {
+    std::vector<uint8_t> val = std::vector<uint8_t>(
+        reinterpret_cast<uint8_t*>(&id),
+        reinterpret_cast<uint8_t*>(&id) + sizeof(id));
```
nit: This is repeated quite a lot throughout this file and we should probably add a function overload for store that just takes a string and performs this conversion internally. This is more on the PT-D team to do.

H-Huang reviewed (commented):
Thanks for sending the PR! I still need to take a closer look at the ProcessGroupUCC implementation, any resources for learning more about UCC would be super helpful. 

One general question, what is the testing strategy going to be? How are we eventually going to upstream the test? Also besides the Torch-UCC test you mentioned in the description, how are we going to test this integration? I believe right now if you tried to do `init_process_group('ucc', ....)` it wouldn't work completely. Perhaps as part of this PR we should at least add a new test file which includes `init_process_group('ucc', ....)` which can be run locally.

vtlam commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2982,6 +3030,7 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
         if pg != GroupMember.NON_GROUP_MEMBER and get_backend(pg) in [
             Backend.GLOO,
             Backend.NCCL,
+            Backend.UCC,
```
I'll remove it.

vtlam commented on CMakeLists.txt:
```diff
@@ -308,6 +308,14 @@ option(USE_DISTRIBUTED "Use distributed" ON)
 cmake_dependent_option(
     USE_MPI "Use MPI for Caffe2. Only available if USE_DISTRIBUTED is on." ON
     "USE_DISTRIBUTED" OFF)
+cmake_dependent_option(
+    USE_UCC "Use UCC. Only available if USE_DISTRIBUTED is on." OFF
+    "USE_DISTRIBUTED" OFF)
+cmake_dependent_option(
+    USE_SYSTEM_UCC "Use system-wide UCC" OFF
+    "USE_UCC" OFF)
```
Yep, the plan is to enable UCC gradually. As of this PR, everything related to UCC is default to OFF, so that it won't break any existing things. Once we port over the tests and harden them, we can start to gradually switch the UCC knobs to default ON.

vtlam commented on torch/distributed/distributed_c10d.py:
```diff
@@ -767,6 +788,33 @@ def _new_process_group_helper(
+            pg = ProcessGroupUCC(prefix_store, rank, world_size, timeout=timeout)
+            # In debug mode and if GLOO is available, wrap in a wrapper PG that
+            # enables enhanced collective checking for debugability.
+            if get_debug_level() == DebugLevel.DETAIL:
+                if not _GLOO_AVAILABLE:
+                    logger.info(
+                        """TORCH_DISTRIBUTED_DEBUG was set to DETAIL, but
+                                GLOO is not available. Build with Gloo to
+                                create a wrapper process group in debug mode
+                                to aid collective desynchronization debugging."""
+                    )
+                else:
+                    pg = _create_process_group_wrapper(
+                        wrapped_pg=pg,
+                        store_prefix=group_name,
+                        store=store,
+                        rank=rank,
+                        world_size=world_size,
+                        timeout=timeout,
+                    )
+            _pg_map[pg] = (Backend.UCC, store)
+            _pg_names[pg] = group_name
```
As mentioned in the other comment, UCC-related flags are OFF in this PR so that it is essentially no-op to existing Pytorch users.
The UCC PG logic is taken from the torch-ucc repo snapshot that passes the tests there.
We have an existing set of UCC tests (as Pytorch PG plugin) that we will port over immediately after this. 
There is also a prerequisite to add ucc/ucx repo to pytorch/third-party.

vtlam commented on torch/distributed/distributed_c10d.py:
```diff
@@ -767,6 +788,33 @@ def _new_process_group_helper(
+            pg = ProcessGroupUCC(prefix_store, rank, world_size, timeout=timeout)
+            # In debug mode and if GLOO is available, wrap in a wrapper PG that
+            # enables enhanced collective checking for debugability.
+            if get_debug_level() == DebugLevel.DETAIL:
```
Currently certain UCC use cases may just rely on the underlying NCCL backend, so i just wanted to the feature on par with NCCL PG. Longer term, we need to revisit this PG wrapping (e.g., with the evolution of Gloo etc).

vtlam commented:
> Thanks for sending the PR! I still need to take a closer look at the ProcessGroupUCC implementation, any resources for learning more about UCC would be super helpful.
> 
> One general question, what is the testing strategy going to be? How are we eventually going to upstream the test? Also besides the Torch-UCC test you mentioned in the description, how are we going to test this integration? I believe right now if you tried to do `init_process_group('ucc', ....)` it wouldn't work completely. Perhaps as part of this PR we should at least add a new test file which includes `init_process_group('ucc', ....)` which can be run locally.

kwen2501 reviewed (approved):
Accepting this PR per internal reviews and upcoming plan of adding tests and features.

Commit:
Native UCC process group for Pytorch (#79918)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/79918

This diff integrates UCC process group as a native component of Pytorch Distributed core. It is based on the existing torch-ucc (https://github.com/facebookresearch/torch_ucc) as the wrapper for UCC collective communication library.
The environment and cmake variables are named in mirroring to the existing process groups such as NCCL and Gloo. Specifically,
- USE_UCC: enables UCC PG. This defaults to OFF, so there is no breakage of existing builds that do not have UCX/UCC external libraries.
- USE_SYSTEM_UCC: uses external UCX and UCC shared libraries that are set accordingly with UCX_HOME and UCC_HOME.

Currently, this diff only supports USE_SYSTEM_UCC=ON, i.e., requiring users to specify external libraries for UCX and UCC. In subsequent diffs, we will add UCX and UCC repos as third-party dependencies in pytorch/third-party.

Test Plan:
Passed Torch-UCC tests that invoke UCC process group. For example:

$ sh test/start_test.sh test/torch_allreduce_test.py --backend gloo --use-cuda
...
Test allreduce: succeeded

Reviewed By: kingchc, kwen2501

Differential Revision: D36973688

fbshipit-source-id: 08b0c05f7cedc9526b53abf0aef3db1bc88b9a75
